### PR TITLE
Wire Signals library

### PR DESCRIPTION
**Pull request by makingthematrix**
This PR replaces the signals functionality from Wire Android codebase with signals from the separate library: https://github.com/wireapp/wire-signals

1. Signals, event streams, `CancellableFuture`, and classes aiding them, were moved to the library.
2. The `Threading` class was split in two. The basic functionality is in the library, the parts specific to Wire Android (e.g. the `Ui` thread) is still in Wire Android. One of the next steps in developing wire-signals aims to make the library part easier to use, so that the Wire Android part can be simplified and then use as an example how wire-signals can be used in other projects.
3. You might have noticed that we have flaky unit tests which regularly force us to write comments "build again" on GitHub.
While working on wire-signals, I was able to de-flake some of them, the ones testing `EventProcessingQueue`. The queue is supposed to turn all thrown exceptions into `Future.failed`, but it doesn't - and now I'm pretty sure it never did. The tests passed from time to time because they finished before the exception was thrown (in the fashion Gizem discovered in other unit tests some time ago).

#### APK
[Download build #2266](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2266/artifact/build/artifact/wire-dev-PR2901-2266.apk)
[Download build #2270](http://10.10.124.11:8080/job/Pull%20Request%20Builder/2270/artifact/build/artifact/wire-dev-PR2901-2270.apk)

### CODE REVIEW

File: wire-android-sync-engine/zmessaging/build.gradle [hunk 1]
```diff
@@ -95,6 +95,7 @@ dependencies {
     implementation "com.wire:icu4j-shrunk:57.1"
     implementation "com.googlecode.mp4parser:isoparser:1.1.7"
     implementation "com.github.joshjdevl.libsodiumjni:libsodium-jni-aar:2.0.2"
+    implementation "com.wire:wire-signals_${LegacyDependencies.SCALA_MAJOR_VERSION}:${Versions.WIRE_SIGNALS}"
```
Comment (gizemb) [inline wire-android-sync-engine/zmessaging/build.gradle:98]: can we use `BuildDependencies.wireSignals` ?

Comment (android10) [reply wire-android-sync-engine/zmessaging/build.gradle:98]: Great Point Gizem. I did not see this one :heart:

Comment (makingthematrix) [reply wire-android-sync-engine/zmessaging/build.gradle]: done

File: buildSrc/src/main/kotlin/Dependencies.kt [hunk 2]
```diff
@@ -137,7 +134,7 @@ object BuildDependencies {
     val libPhoneNumber = "com.googlecode.libphonenumber:libphonenumber:${Versions.LIB_PHONE_NUMBER}"
     val pinEditText = "com.poovam:pin-edittext-field:${Versions.PIN_EDITTEXT}"
 
-    val wireSignals = "com.wire:wire-signals_${LegacyDependencies.SCALA_MAJOR_VERSION}:${Versions.WIRE_SIGNALS}"
+    val wireSignals = "com.wire:wire-signals_${LegacyDependencies.SCALA_MAJOR_VERSION}:${LegacyDependencies.WIRE_SIGNALS}"
```
Comment (gizemb) [inline buildSrc/src/main/kotlin/Dependencies.kt:137]: hmm, this can be directly moved into LegacyDependencies. Not so important though.

**Pull request merged**